### ISSUE TITLE
"Standard Mode: Invalid DPMI return from 0367:5F7A" when installing Windows 98SE

### ISSUE THREAD

**Issue by someperson** [joncampbell123/dosbox-x#1743]
**Describe the bug**
I'm getting the error "Standard Mode: Invalid DPMI return from 0367:5F7A" when attempting to install Windows 98SE.

**To Reproduce**
1. imgmake path\to\disk.vhd -t hd -size 65536 -fat 32
2. imgmount C path\to\disk.vhd
3. imgmount D path\to\win98se.iso
4. xcopy D:\WIN98 C:\WIN98 /I/E
5. C:
6. cd WIN98
7. setup
8. Once I get past the initial wizard where I pick components, etc., the installation crashes with this error.

**Expected behavior**
Windows 98SE installs correctly.

**Screenshots**
![error](https://user-images.githubusercontent.com/376450/87970407-4bd80500-ca89-11ea-9bb6-c9690cc232e5.png)

**Environment (please complete the following information):**
 - Windows 10
 - DOSBox-X 0.83.3
 - Configuration: [dosbox-x.conf.txt](https://github.com/joncampbell123/dosbox-x/files/4949298/dosbox-x.conf.txt)

**Comment by joncampbell123** [comment]
Could the 65536MB size have anything to do with it?

Try larger or smaller sizes.

Also try playing with the -spc option in IMGMAKE to control sectors per cluster i.e. -spc 16, -spc 32, etc. There are known issues with Windows 98's installer if the FAT table is too large for some internal table.

**Comment by rderooy** [comment]
I have been able to install Win98SE on HD images up to 128GB, although that
did cause a warning during installation. But the installation succeeded.
 I did not get this DPMI error. How much memory did you give Win98SE?
And did you follow the wiki, in particular did you use the provided config
file?
https://github.com/joncampbell123/dosbox-x/wiki/Guide%3AInstalling-Windows-98

**Comment by someperson** [comment]
I did use the provided config file, the only difference I can think of offhand being that I enabled a 2x scaler. I just tried again this time using 32768 MB for the size. It got a lot further, but I ran into this error during the installation:
![error2](https://user-images.githubusercontent.com/376450/87982587-a169dd00-ca9c-11ea-8fb4-5379ad3b2d9d.png)

**Comment by rderooy** [comment]
Hmm...  GPF's are normally triggered by using the wrong core= value. You're
sure you're using core=normal as in my reference config file on the wiki?

**Comment by someperson** [comment]
I definitely have core=normal set in my configuration. I was running in turbo mode during this part of the installation though. Could that have caused this?

**Comment by Wengier** [comment]
@rderooy I just tried myself to install Windows 98SE using the steps he mentioned (except that the image I created was 4GB FAT32) on a Windows 10 machine, and I also used the exact same config file as you provided in the Guide:Installing Windows 98 Wiki page, and I can indeed confirm the error (I did not enable Turbo mode). The GPF error occurs in the very end of the Setup program (after the progress bar reaches 100%), but it is not possible to start Windows 98SE in such case because the installation still did not finish. At this time I am not sure whether this issue is Windows-specific or not though (apparently he is also using it in a Windows 10 machine).

**Comment by Wengier** [comment]
The main problem to find out the cause of the GPF is that the error only occurs in the very end of the Setup, which means you have to wait a very long time before the error shows up.

**Comment by rderooy** [comment]
Strange, because I did quite a few installs for testing, and did not encounter either the GPF or the DMPI error. But I did it with the Linux SDL1 version.

**Comment by Wengier** [comment]
@rderooy Perhaps you can try the installation in the Windows version too? I know that most stuff should work exactly the same cross-platform, but there are always exceptions due to things like compilers etc.

**Comment by rderooy** [comment]
I am traveling this week, so I will not be able to try it until the weekend
at the earliest.

**Comment by Wengier** [comment]
I just tried again with the same configuration, but this time with a 2GB FAT32, and the Windows 98SE installation was a success on my Windows 10 system. At this time I am not very sure, but it is still possible that the error has something to do with the size or geometry of the disk image.

**Comment by rderooy** [comment]
Another thing to note, initially my installs of Win98 where "Compact" installs, simply to cut down on the time it takes to do the installation. I also left everything to default, so default keyboard, country, etc. And in the case of the printer just skipped it.

With my later testing I used a MSBATCH.INF template file to answer the questions, such that I could let it run unattended. But once again, used minimal installs.

The majority of testing was with the Linux SDL1 version, but I did do a couple of Linux SDL2 installs, and did not encounter any problems with it.

In any case, I just did a test with a 2GB FAT32 C: Drive on Windows 10. I did a Custom install, with almost every package selected, and I got the DPMI error.

**Comment by rderooy** [comment]
Just did a Compact install with a 2GB FAT32 drive, and got the same DPMI error. These are the message from the LOG the moment it happened:
```
LOG: pixratio 1.350, dw false, dh false
LOG: Aspect ratio: 720 x 540  xToY=1.333 yToX=0.750
LOG: Screen report: Method 'Win98base' (2560.000 x 1440.000 pixels) at (0.000 x 0.000) (541.867 x 304.800 mm) (21.333 x 12.000 in) (120.000 x 120.000 DPI)
LOG: Screen report: Method 'Win98base' (2560.000 x 1440.000 pixels) at (0.000 x 0.000) (541.867 x 304.800 mm) (21.333 x 12.000 in) (120.000 x 120.000 DPI)
LOG: Screen report: Method 'Win98base' (2560.000 x 1440.000 pixels) at (0.000 x 0.000) (541.867 x 304.800 mm) (21.333 x 12.000 in) (120.000 x 120.000 DPI)
LOG: DEBUG: INT 2Fh Windows 286/386 DOSX exit broadcast issued (DX=0x0001)
LOG:  [286 DOS extender]
LOG:
LOG: 1673069657 ERROR DOSMISC:DOS:INT 2F Unhandled call AX=1606
LOG: Aspect ratio: 720 x 540  xToY=1.333 yToX=0.750
LOG: Screen report: Method 'Win98base' (2560.000 x 1440.000 pixels) at (0.000 x 0.000) (541.867 x 304.800 mm) (21.333 x 12.000 in) (120.000 x 120.000 DPI)
```

**Comment by Wengier** [comment]
It is possible that the either error is non-deterministic on the Windows machine, i.e. not always happen, but with a good chance to happen (>50%). But this of course means the installation is not reliable and there are bugs needs to be fixed.

**Comment by rderooy** [comment]
It seems to crash back to the DOS prompt pretty reliably here :-)

To ensure it had nothing to do with Turbo mode, I tried to do another install from scratch without touching the Turbo mode, and it again crashed back to the DOS prompt with the DPMI error. Do it seems Turbo mode has no effect either way.

**Comment by Wengier** [comment]
@rderooy I think I did see the DPMI error myself when I tried to install Win98SE too, but it disappeared when I restarted the Setup (without creating a new disk image). I tried with the latest development code of 0.83.4 at that time. The Windows installer package on the other hand includes the official DOSBox-X 0.83.3 Windows binaries released by @joncampbell123 intact.

**Comment by pancirno** [comment]
I confirm, the DPMI error appears on the first time you run the setup without restarting dosbox-x, on the next launch the DPMI error didn't appear.

**Comment by rderooy** [comment]
@pancirno I can indeed confirm that when you get the DPMI error, that if you just run SETUP again without restarting DOSBox-X that it does go further. But restarting DOSBox-X (with the same, unmodified, HDD image), and you get the DPMI error again. So the state of the virtual machine is not the same any more and it succeeds for some reason.

I can also confirm the GPF here at the end of the 1st install phase, when the file copy is at 100%.

**Comment by rderooy** [comment]
A few updates.

- The DPMI issue also happens on Linux
- It does not happen however, when I use my MSBATCH.INF answer file, which is why I did not notice it.

**Comment by rderooy** [comment]
Without using an MSBATCH.INF file...

Success:
- DOS VER 5.0 with FAT16 2GB drive

Fails with DPMI error:
- DOS VER 7.1 with FAT16 2GB drive
- DOS VER 7.1 with FAT32 2GB drive

**Comment by pancirno** [comment]
After switching to Dos 5.0 I managed to install and boot into Windows 98 SE

**Comment by rderooy** [comment]
Here is my MSBATCH.INF just fix its name it and place it in the directory next to SETUP.EXE and Win98SE will install in cases where it otherwise fails with the DPMI error.

[MSBATCH.INF.txt](https://github.com/joncampbell123/dosbox-x/files/4977980/MSBATCH.INF.txt)

This batch file also will try to automatically configure the NE2000, so if you don't have NE2000 enabled in your dosbox-x config file, you may want to comment it out, or delete the references to PNP80D6.

**Comment by Wengier** [comment]
@rderooy and @pancirno Thanks for the experiments. I think I have a fix for the DPMI error. The installation is still ongoing, so I am still waiting to see if the GPF error will occur in the end.

**Comment by Wengier** [comment]
I have already submitted the fix in my latest pull request. With the new code, I tried Windows 98SE installations four times using the exact same config file as provided in the Wiki without using a batch file, and the DPMI error no longer occurred. Out of the four Windows 98SE installations I made, I have allowed the Setup to proceed to the end 3 times, and the GPF error also did not occur when the progress bar reached 100%, so it very likely also fixed the GPF error in the end.
